### PR TITLE
feat: use pino for logging

### PR DESCRIPTION
**Pull request by mattyod**
* using `server.logger` and `request.logger` in almost all scenarios
* allowing errors to be thrown, so that the user gets feedback that something is wrong rather than presenting the view as if there where no results available
* refactor of tests to check when endpoints actually error

### CODE REVIEW

File: app/api/applications.js [hunk 1]
```diff
@@ -28,36 +25,15 @@ async function getApplications (searchType, searchText, limit, offset, filterSta
...
+    return payload
   } catch (err) {
-    console.log(err)
-    return { applications: [], total: 0, applicationStatus: [] }
+    logger.setBindings({ err, endpoint })
+    throw err
   }
 }
 
-async function withdrawApplication (reference, user, status) {
```
Comment (aaroncarroll82) [inline app/api/applications.js:42]: 👏 Removing unused stuff, might be a first for this application, but good to see!

File: app/api/applications.js [hunk 2]
```diff
@@ -86,51 +62,39 @@ async function updateApplicationStatus (reference, user, status) {
...
-      throw new Error(`HTTP ${response.res.statusCode} (${response.res.statusMessage})`)
-    }
-    return response.payload
+    const { payload } = await wreck.get(endpoint, { json: true })
+    return payload
   } catch (err) {
-    console.error(`Getting application events for ${reference} failed: ${err.message}`)
-
-    return { eventRecords: [] }
+    logger.setBindings({ err })
```
Comment (aaroncarroll82) [inline app/api/applications.js:90]: Could probably add endpoint here matching the others above

File: app/api/gov-holiday.js [hunk 3]
```diff
@@ -1,42 +1,34 @@
...
-    if (!isHoliday) isHoliday = await module.exports.isTodayCustomHoliday()
-    return isHoliday
-  } catch (err) {
-    console.error(`Checking holiday failed: ${err.message}`)
-    return false // Assume not a holiday in case of error
-  }
+  const holidays = await module.exports.getHolidayCalendarForEngland()
+  const today = new Date().toISOString().split('T')[0] // Format today's date as YYYY-MM-DD
+  let isHoliday = holidays?.some(holiday => holiday.date === today)
+  if (!isHoliday) isHoliday = await module.exports.isTodayCustomHoliday()
```
Comment (aaroncarroll82) [inline app/api/gov-holiday.js:30]: I know you didn't write this line, was already here but I find this really confusing. Would prefer an if statement using curly braces as per others, but also just the repeated isHoliday makes it hard to parse `if (!isHoliday) isHoliday`. I had to triple take on that one

File: app/constants/application-urls.js [hunk 4]
```diff
@@ -0,0 +1,7 @@
+const { applicationApiUri } = require('../config')
+
+module.exports.APPLICATION = `${applicationApiUri}/application`
+module.exports.APPLICATION_CLAIM = `${applicationApiUri}/application/claim`
+module.exports.APPLICATION_GET = `${applicationApiUri}/application/get`
+module.exports.APPLICATION_SEARCH = `${applicationApiUri}/application/search`
+module.exports.CLAIM_BY_REFERENCE = `${applicationApiUri}/claim/get-by-reference`
```
Comment (aaroncarroll82) [inline app/constants/application-urls.js:7]: Is it worth adding all the URLs defined in the api files above into this file so they're in one place?

File: app/routes/recommend-to-reject.js [hunk 5]
```diff
@@ -31,27 +30,23 @@ module.exports = {
...
-          return h.redirect(`/view-claim/${request.payload.reference}${request.payload?.returnPage && '?returnPage=' + request.payload?.returnPage}`)
-        } else {
-          return h.redirect(`/view-agreement/${request.payload.reference}?page=${request.payload.page}`)
-        }
-      } catch (error) {
-        console.error(`routes:recommend-to-reject: Error when processing request: ${error.message}`)
-        throw Boom.internal(error.message)
+      const userRole = mapAuth(request)
+      request.logger.setBindings({ userRole })
+      if (!userRole.isRecommender && !userRole.isAdministrator) {
```
Comment (aaroncarroll82) [inline app/routes/recommend-to-reject.js:35]: Might be slightly beyond the scope here, but I'm seeing this repeated a couple of times, maybe worth creating a util function somewhere along the lines of `userAllowedToRecommend(userRole) : boolean`

File: package.json [hunk 6]
```diff
@@ -37,7 +37,7 @@
     "@ministryofjustice/frontend": "^2.1.3",
```
Comment (aaroncarroll82) [inline package.json:37]: Not specific to this line, but remember to increase version number of the app in package.json

File: app/constants/application-urls.js (hunk 4 above)
Comment (mattyod) [reply app/constants/application-urls.js:7]: I added this file at the beginning when I was taking a slightly different approach an moving the calls to the application api up into their handlers. Realised that this would have a significant impact on our testing so (for now at least) I've left them in their api folders and passed the logger down to them. So think it's reasonable to just leave the single use strings in the api call files.

TL;DR: I intend to remove this file.

File: app/api/gov-holiday.js (hunk 3 above)
Comment (mattyod) [reply app/api/gov-holiday.js:30]: I've been trying to avoid too much 'off topic' refactoring, which has been hard 😂, but as this line now has my name on it I'll make an exception 😉

File: app/routes/recommend-to-reject.js (hunk 5 above)
Comment (mattyod) [reply app/routes/recommend-to-reject.js:35]: Personally feel that's an abstraction too far, as the userRole object that's created gets checked in subtly different ways depending on the situation. 

Possibly the generation of the userRole object is something that could be done as part of the request lifecycle an made available on `request.app` for every request though. One to discuss outside the scope of this PR?

Screenshot 2024-11-13 at 11 31 46 am

Comment (mattyod) [reply app/routes/recommend-to-reject.js:35]: https://hapi.dev/api/?v=21.3.12#-requestapp

File: app/crons/process-on-hold/process.js [hunk 7]
```diff
@@ -7,47 +7,29 @@ const getCommonData = () => {
...
-    } else {
-      console.log(`${new Date().toISOString()} processing OnHold applications:Nothing to process today`)
+const processOnHoldApplications = async (logger) => {
+  const { searchType, searchText, datePast24Hrs } = getCommonData()
+  const apps = await getApplications(searchType, searchText, 9999, 0, [], { field: 'CREATEDAT', direction: 'ASC' }, logger)
+  logger.setBindings({ applicationsTotal: apps.total })
+  if (apps.total > 0) {
+    const applicationRefs = apps.applications.filter(a => new Date(a.updatedAt) <= datePast24Hrs).map(a => a.reference)
+    logger.setBindings({ applicationRefs })
+    for (const appRef of applicationRefs) {
```
Comment (RobCatton) [inline app/crons/process-on-hold/process.js:17]: How many applicationRefs can we be looping through here? Just asking because having an awaited function inside a for loop could end up being _**real**_ slow, is there a chance we should be using await Promise.all here? Or something similiar.

If you think its out of scope, which it very reasonably could be, then maybe just needs flagging to revisit.

File: app/crons/process-on-hold/process.js [hunk 8]
```diff
@@ -7,47 +7,29 @@ const getCommonData = () => {
...
-    } else {
-      console.log(`${new Date().toISOString()} - processing OnHold claims: Nothing to process today`)
+const processOnHoldClaims = async (logger) => {
+  const { searchType, searchText, datePast24Hrs } = getCommonData()
+  const { claims, total } = await getClaims(searchType, searchText)
+  logger.setBindings({ claimsTotal: total })
+  if (total > 0) {
+    const claimRefs = claims.filter(a => new Date(a.updatedAt) <= datePast24Hrs).map(a => a.reference)
+    logger.setBindings({ claimRefs })
+    for (const claimRef of claimRefs) {
```
Comment (RobCatton) [inline app/crons/process-on-hold/process.js:29]: Same question as above

File: test/unit/routes/utils/regex-checker.test.js [hunk 9]
```diff
@@ -2,55 +2,43 @@
 const regexChecker = require('../../../../app/routes/utils/regex-checker')
 
 describe('regexChecker', () => {
```
Comment (RobCatton) [inline test/unit/routes/utils/regex-checker.test.js:4]: Do we need a regex checker? I thought that was built into JS.

File: app/crons/process-on-hold/process.js (hunk 7 above)
Comment (mattyod) [reply app/crons/process-on-hold/process.js:17]: Yeah fair call, it could be a few (however many happen in a day). we could get the refs and promise all or even just remove the await as we expect no return values, it seems to be just fire and forget.

Although as it's a daily cron job, not sure how much the speed of execution really matters.

File: test/unit/routes/utils/regex-checker.test.js (hunk 9 above)
Comment (mattyod) [reply test/unit/routes/utils/regex-checker.test.js:4]: It's a kind of misleadingly named function, it's more a 'run regex safely' method but personally I'd just get rid of it, use the regular expressions inline where they are needed and make sure non of them are invalid. I actually have a note for this a small PR for later.

File: app/routes/recommend-to-reject.js (hunk 5 above)
Comment (aaroncarroll82) [reply app/routes/recommend-to-reject.js:35]: We've got some work coming up to expand the functionality for changing statuses, might be good to have a look at this when that work is done

File: app/crons/process-on-hold/process.js (hunk 7 above)
Comment (RobCatton) [reply app/crons/process-on-hold/process.js:17]: Thats fair yeah - lets not spend time trying to fix something that might not need fixing.

File: test/unit/routes/utils/regex-checker.test.js (hunk 9 above)
Comment (RobCatton) [reply test/unit/routes/utils/regex-checker.test.js:4]: Yeah I think I prefer that too. Cool, if its in another PR then we can do it there.

**Review by RobCatton — APPROVED** [review]
